src/utils/str_utils.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "logger.h"
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+
+/**
+ * @brief Calculate the length of data that would be written by snprintf,
+ *       ensuring it does not exceed buffer size.
+ *
+ * @param dest Pointer to the destination buffer.
+ * @param data_size Size of the destination buffer.
+ * @param fmt Format string.
+ * @param ... Additional arguments for the format string.
+ * @return Length of data that would be written, including null terminator,
+ *         or negative value on encoding error.
+ */
+inline static int snprintf_len(char *dest, size_t data_size, const char *fmt,
```

Comment:
This feels like it can easily go wrong, let's add some tests to catch this! Please add a `/src/utils/test` subdir and have a `snprintf_len.c` test script!